src/nni_engine.cpp

```diff
@@ -33,43 +33,88 @@ void NNIEngine::RunInit() {
+  UpdateAdjacentNNIs();
+  // (5) Reset Accepted NNIs and GraftDAG and save results.
+  UpdateAcceptedNNIs();
+  UpdateRejectedNNIs();
+}
+
+// ** Static Filter Functions
+
+void NNIEngine::NoInit(NNIEngine &this_nni_engine, GPEngine &this_gp_engine,
+                       GraftDAG &this_graft_dag) {}
+
+void NNIEngine::NoUpdate(NNIEngine &this_nni_engine, GPEngine &this_gp_engine,
+                         GraftDAG &this_graft_dag) {}
+
+double NNIEngine::GetBestPreNNILikelihood(const NNIOperation &nni) {
+  size_t prenni_edge_idx = NoId;
+  double prenni_likelihood = -INFINITY;
+  SizeVector prenni_idxs;
+  DoubleVector prenni_likelihoods;
+  for (const SubsplitClade clade_to_swap :
+       {SubsplitClade::Right, SubsplitClade::Left}) {
+    NNIOperation pre_nni = nni.NNIOperationFromNeighboringSubsplits(
+        (clade_to_swap == SubsplitClade::Right));
+    if (dag_.ContainsNode(pre_nni.GetParent()) &&
+        dag_.ContainsNode(pre_nni.GetChild())) {
+      const auto &parent_id = dag_.GetDAGNodeId(pre_nni.GetParent());
+      const auto &child_id = dag_.GetDAGNodeId(pre_nni.GetChild());
+      if (dag_.ContainsEdge(parent_id, child_id)) {
+        const auto &edge_idx = dag_.GetEdgeIdx(parent_id, child_id);
+        const auto likelihood = GetDAGEdgeLikelihood(edge_idx);
+        if (prenni_likelihood < likelihood) {
+          prenni_likelihood = likelihood;
+          prenni_edge_idx = edge_idx;
+        }
+        prenni_idxs.push_back(edge_idx);
+        prenni_likelihoods.push_back(likelihood);
+      }
+    }
+  }
+  Assert(prenni_edge_idx != NoId,
+         "Pre-NNI could not be found in DAG based on Post-NNI.");
+  return prenni_likelihood;
 }
 
-// ** Filter Functions
+double NNIEngine::GetPostNNILikelihood(const NNIOperation &nni) {
+  return scored_nnis_[nni];
+}
 
-double NNIEngine::NoEvaluate(NNIEngine &this_nni_engine, GPEngine &this_gp_engine,
-                             GraftDAG &this_graft_dag, const NNIOperation &nni) {
-  return 0.0;
+void NNIEngine::UpdateScoreAdjacentNNIsByLikelihood(NNIEngine &this_nni_engine,
+                                                    GPEngine &this_gp_engine,
+                                                    GraftDAG &this_graft_dag) {
+  this_nni_engine.ScoreAdjacentNNIsByLikelihood();
 }
 
-bool NNIEngine::NoFilter(NNIEngine &this_nni_engine, GPEngine &this_gp_engine,
-                         GraftDAG &this_graft_dag, const NNIOperation &nni) {
-  return true;
+bool NNIEngine::FilterPostNNIBetterThanPreNNI(NNIEngine &this_nni_engine,
+                                              GPEngine &this_gp_engine,
+                                              GraftDAG &this_graft_dag,
+                                              const NNIOperation &nni) {
+  const double postnni_likelihood = this_nni_engine.GetPostNNILikelihood(nni);
+  const double prenni_likelihood = this_nni_engine.GetBestPreNNILikelihood(nni);
+  const double diff = postnni_likelihood - prenni_likelihood;
```

Comment:
Wouldn't this be clearer and simpler if we just had 
```
return postnni_likelihood > prenni_likelihood;
```
?

Reply:
Yep!  Changed.